libdrgn/program.c

```diff
@@ -788,20 +821,146 @@ struct drgn_error *drgn_program_find_prstatus_by_tid(struct drgn_program *prog,
 						     struct nstring *ret)
 {
 	struct drgn_error *err;
-	struct drgn_prstatus_map_iterator it;
+	struct drgn_thread *thread;
+	err = drgn_program_find_thread(prog, tid, &thread);
+	if (err)
+		return err;
+
+	if (!thread) {
+		ret->str = NULL;
+		ret->len = 0;
+		return NULL;
+	}
+	*ret = thread->prstatus;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_find_thread(struct drgn_program *prog, uint32_t tid,
+			 struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+	struct drgn_thread_set_iterator it;
 
 	assert(!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL));
-	err = drgn_program_cache_prstatus(prog);
+	err = drgn_program_cache_threads(prog);
 	if (err)
 		return err;
 
-	it = drgn_prstatus_map_search(&prog->prstatus_map, &tid);
+	it = drgn_thread_set_search(&prog->thread_set, &tid);
 	if (!it.entry) {
-		ret->str = NULL;
-		ret->len = 0;
+		*ret = NULL;
 		return NULL;
 	}
-	*ret = it.entry->value;
+	*ret = it.entry;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_program_crashed_thread(struct drgn_program *prog, struct drgn_thread **ret)
+{
+	struct drgn_error *err;
+
+	assert(!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL));
+	if (prog->flags & DRGN_PROGRAM_IS_LIVE)
+		return drgn_error_create(
+			DRGN_ERROR_INVALID_ARGUMENT,
+			"the crashed thread can only be found when debugging a core dump");
+	err = drgn_program_cache_threads(prog);
+	if (err)
+		return err;
+	if (!prog->crashed_thread)
+		return drgn_error_create(DRGN_ERROR_OTHER, "crashed thread not found");
+	*ret = prog->crashed_thread;
+	return NULL;
+}
+
+LIBDRGN_PUBLIC struct drgn_error *
+drgn_thread_iterator_create(struct drgn_program *prog,
+			    struct drgn_thread_iterator **ret)
+{
+	struct drgn_error *err;
+	*ret = malloc(sizeof(**ret));
+	if (!*ret)
+		return &drgn_enomem;
+	(*ret)->prog_is_live = prog->flags & DRGN_PROGRAM_IS_LIVE;
+	if ((*ret)->prog_is_live) {
+		if (!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL)) {
+			err = drgn_error_create(
+				DRGN_ERROR_INVALID_ARGUMENT,
+				"iterating over the threads of a live userspace process is not currently supported");
+			goto out;
+		}
+		drgn_object_init(&(*ret)->entry.thread, prog);
```

Comment:
`(*ret)->entry.thread` should be deinitialized if there is an error

Reply:
Actually this is unnecessary because it isn't populated with any value, just inited.